### PR TITLE
FEATURE: SchemaExtender

### PR DESCRIPTION
**Pull request by Torsten85**
Port of the original [extendSchema](https://github.com/graphql/graphql-js/blob/master/src/utilities/extendSchema.js) implementation

### CODE REVIEW

**Comment by vladar** [comment]
Cool. But looks like some tests are failing.

**Comment by simPod** [comment]
_BTW, you can fix a lot of CS violation by simply running `vendor/bin/phpcbf`_

File: tests/Utils/SchemaExtenderTest.php [hunk 1]
```diff
@@ -0,0 +1,1924 @@
+<?php
+
+declare(strict_types=1);
+
+namespace GraphQL\Tests;
```
Comment (simPod) [inline tests/Utils/SchemaExtenderTest.php]: GraphQL\Tests\Utils

Comment (Torsten85) [reply tests/Utils/SchemaExtenderTest.php]: Thanks!

File: src/Utils/SchemaExtender.php [hunk 2]
```diff
@@ -0,0 +1,627 @@
...
+
+    /** @var ASTDefinitionBuilder */
+    protected static $astBuilder;
+
+    /**
+     * @return TypeExtensionNode[]|null
+     */
+    protected static function getExtensionASTNodes(NamedType $type) : ?array
+    {
+        if (! isset($type->name)) {
```
Comment (Torsten85) [inline src/Utils/SchemaExtender.php]: This line generates the error
```
Accessing name on the interface GraphQL\Type\Definition\NamedType suggest that you code against a concrete implementation. How about adding an instanceof check?
```
How should I handle this? There seems to be no interface that provides the "name" property

**Comment by Torsten85** [comment]
Ok there is nothing more for me to do now. Scrutinizer declares some method / classes as too complex, but they match the reference implementation.

File: src/Validator/DocumentValidator.php [hunk 3]
```diff
@@ -273,4 +298,19 @@ public static function isValidLiteralValue(Type $type, $valueNode)
...
         return $context->getErrors();
     }
+
+    public static function assertValidSDLExtension(DocumentNode $documentAST, Schema $schema)
+    {
+        $errors = self::visitUsingRules($schema, new TypeInfo($schema), $documentAST, self::sdlRules());
+        if (count($errors) !== 0) {
+            throw new Error(
+                implode(
+                    '\n\n',
```
Comment (simPod) [inline src/Validator/DocumentValidator.php]: not sure whats better, personally I'd use PHP_EOL constants twice (`PHP_EOL.PHP_EOL`)

File: src/Validator/DocumentValidator.php [hunk 4]
```diff
@@ -273,4 +298,19 @@ public static function isValidLiteralValue(Type $type, $valueNode)
...
+                    array_map(static function (Error $error) {
```
Comment (simPod) [inline src/Validator/DocumentValidator.php]: here we're missing return type, is it possible to add?

File: src/Validator/Rules/KnownArgumentNamesOnDirectives.php [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+
+    public function getVisitor(ValidationContext $context)
+    {
+        $directiveArgs     = [];
+        $schema            = $context->getSchema();
+        $definedDirectives = $schema !== null ? $schema->getDirectives() : Directive::getInternalDirectives();
+
+        foreach ($definedDirectives as $directive) {
+            $directiveArgs[$directive->name] = array_map(
+                static function ($arg) {
```
Comment (simPod) [inline src/Validator/Rules/KnownArgumentNamesOnDirectives.php]: can we add return type? also, is the type of $arg known/not mixed?

File: src/Validator/Rules/KnownArgumentNamesOnDirectives.php [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+
+            $name = $def->name->value;
+            if ($def->arguments !== null) {
+                $arguments = $def->arguments;
+
+                if ($def->arguments instanceof NodeList) {
+                    $arguments = iterator_to_array($def->arguments->getIterator());
+                }
+
+                $directiveArgs[$name] = array_map(static function ($arg) {
```
Comment (simPod) [inline src/Validator/Rules/KnownArgumentNamesOnDirectives.php]: is it possible to have return type?

File: src/Validator/Rules/ProvidedRequiredArgumentsOnDirectives.php [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+        $requiredArgsMap   = [];
+        $schema            = $context->getSchema();
+        $definedDirectives = $schema->getDirectives();
+
+        foreach ($definedDirectives as $directive) {
+            $requiredArgsMap[$directive->name] = Utils::keyMap(
+                array_filter($directive->args, static function (FieldArgument $arg) {
+                    return $arg->getType() instanceof NonNull && ! isset($arg->defaultValue);
+                }),
+                static function ($arg) {
```
Comment (simPod) [inline src/Validator/Rules/ProvidedRequiredArgumentsOnDirectives.php]: same here

File: src/Validator/Rules/ProvidedRequiredArgumentsOnDirectives.php [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+
+    public function getVisitor(ValidationContext $context)
+    {
+        $requiredArgsMap   = [];
+        $schema            = $context->getSchema();
+        $definedDirectives = $schema->getDirectives();
+
+        foreach ($definedDirectives as $directive) {
+            $requiredArgsMap[$directive->name] = Utils::keyMap(
+                array_filter($directive->args, static function (FieldArgument $arg) {
```
Comment (simPod) [inline src/Validator/Rules/ProvidedRequiredArgumentsOnDirectives.php]: same here

File: src/Validator/Rules/ProvidedRequiredArgumentsOnDirectives.php [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+
+            $requiredArgsMap[$def->name->value] = Utils::keyMap(
+                $arguments ? array_filter($arguments, static function (Node $argument) {
+                    return $argument instanceof NonNullTypeNode &&
+                        (
+                            ! isset($argument->defaultValue) ||
+                            $argument->defaultValue === null
+                        );
+                }) : [],
+                static function (NamedTypeNode $argument) {
```
Comment (simPod) [inline src/Validator/Rules/ProvidedRequiredArgumentsOnDirectives.php]: same here

File: src/Validator/Rules/ProvidedRequiredArgumentsOnDirectives.php [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+            if (is_array($def->arguments)) {
+                $arguments = $def->arguments;
+            } elseif ($def->arguments instanceof NodeList) {
+                $arguments = iterator_to_array($def->arguments->getIterator());
+            } else {
+                $arguments = null;
+            }
+
+            $requiredArgsMap[$def->name->value] = Utils::keyMap(
+                $arguments ? array_filter($arguments, static function (Node $argument) {
```
Comment (simPod) [inline src/Validator/Rules/ProvidedRequiredArgumentsOnDirectives.php]: same here

File: src/Validator/Rules/ProvidedRequiredArgumentsOnDirectives.php [hunk 11]
```diff
@@ -0,0 +1,109 @@
...
+                $directiveName = $directiveNode->name->value;
+                $requiredArgs  = $requiredArgsMap[$directiveName] ?? null;
+                if (! $requiredArgs) {
+                    return;
+                }
+
+                $argNodes   = $directiveNode->arguments ?: [];
+                $argNodeMap = Utils::keyMap(
+                    $argNodes,
+                    static function ($arg) {
```
Comment (simPod) [inline src/Validator/Rules/ProvidedRequiredArgumentsOnDirectives.php]: same here

File: tests/Utils/SchemaExtenderTest.php [hunk 12]
```diff
@@ -0,0 +1,1929 @@
...
+        $SomeScalarType = new CustomScalarType([
+            'name' => 'SomeScalar',
+            'serialize' => static function ($x) {
+                return $x;
+            },
+        ]);
+
+        $SomeInterfaceType = new InterfaceType([
+            'name' => 'SomeInterface',
+            'fields' => static function () use (&$SomeInterfaceType) {
```
Comment (simPod) [inline tests/Utils/SchemaExtenderTest.php]: same here

File: tests/Utils/SchemaExtenderTest.php [hunk 13]
```diff
@@ -0,0 +1,1929 @@
...
+    /** @var Directive */
+    protected $FooDirective;
+
+    public function setUp()
+    {
+        parent::setUp();
+
+        $SomeScalarType = new CustomScalarType([
+            'name' => 'SomeScalar',
+            'serialize' => static function ($x) {
```
Comment (simPod) [inline tests/Utils/SchemaExtenderTest.php]: same here

File: tests/Utils/SchemaExtenderTest.php [hunk 14]
```diff
@@ -0,0 +1,1929 @@
...
+                    'name' => [ 'type' => Type::string()],
+                    'some' => [ 'type' => $SomeInterfaceType],
+                ];
+            },
+        ]);
+
+        $FooType = new ObjectType([
+            'name' => 'Foo',
+            'interfaces' => [$SomeInterfaceType],
+            'fields' => static function () use ($SomeInterfaceType, &$FooType) {
```
Comment (simPod) [inline tests/Utils/SchemaExtenderTest.php]: same here

File: tests/Utils/SchemaExtenderTest.php [hunk 15]
```diff
@@ -0,0 +1,1929 @@
...
+                    'some' => [ 'type' => $SomeInterfaceType ],
+                    'tree' => [ 'type' => Type::nonNull(Type::listOf($FooType))],
+                ];
+            },
+        ]);
+
+        $BarType = new ObjectType([
+            'name' => 'Bar',
+            'interfaces' => [$SomeInterfaceType],
+            'fields' => static function () use ($SomeInterfaceType, $FooType) {
```
Comment (simPod) [inline tests/Utils/SchemaExtenderTest.php]: same here

File: tests/Utils/SchemaExtenderTest.php [hunk 16]
```diff
@@ -0,0 +1,1929 @@
...
+            'name' => 'SomeEnum',
+            'values' => [
+                'ONE' => [ 'value' => 1 ],
+                'TWO' => [ 'value' => 2 ],
+            ],
+        ]);
+
+        $SomeInputType = new InputObjectType([
+            'name' => 'SomeInput',
+            'fields' => static function () {
```
Comment (simPod) [inline tests/Utils/SchemaExtenderTest.php]: same here

File: tests/Utils/SchemaExtenderTest.php [hunk 17]
```diff
@@ -0,0 +1,1929 @@
...
+                    'name' => [ 'type' => Type::string() ],
+                    'some' => [ 'type' => $SomeInterfaceType ],
+                    'foo' => [ 'type' => $FooType ],
+                ];
+            },
+        ]);
+
+        $BizType = new ObjectType([
+            'name' => 'Biz',
+            'fields' => static function () {
```
Comment (simPod) [inline tests/Utils/SchemaExtenderTest.php]: same here

File: tests/Utils/SchemaExtenderTest.php [hunk 18]
```diff
@@ -0,0 +1,1929 @@
...
+                DirectiveLocation::ENUM_VALUE,
+                DirectiveLocation::INPUT_OBJECT,
+                DirectiveLocation::INPUT_FIELD_DEFINITION,
+            ],
+        ]);
+
+        $this->testSchema = new Schema([
+            'query' => new ObjectType([
+                'name' => 'Query',
+                'fields' => static function () use ($FooType, $SomeScalarType, $SomeUnionType, $SomeEnumType, $SomeInterfaceType, $SomeInputType) {
```
Comment (simPod) [inline tests/Utils/SchemaExtenderTest.php]: same here

File: tests/Utils/SchemaExtenderTest.php [hunk 19]
```diff
@@ -0,0 +1,1929 @@
...
+        $this->assertEquals(SchemaPrinter::doPrint($this->testSchema), $originalPrint);
+        return $extendedSchema;
+    }
+
+    protected function printTestSchemaChanges(Schema $extendedSchema) : string
+    {
+        $ast              = Parser::parse(SchemaPrinter::doPrint($extendedSchema));
+        $ast->definitions = array_values(array_filter(
+            iterator_to_array($ast->definitions->getIterator()),
+            function ($node) {
```
Comment (simPod) [inline tests/Utils/SchemaExtenderTest.php]: same here

File: tests/Utils/SchemaExtenderTest.php [hunk 20]
```diff
@@ -0,0 +1,1929 @@
...
+
+    /**
+     * @param mixed[]|null $options
+     */
+    protected function extendTestSchema(string $sdl, ?array $options = null) : Schema
+    {
+        $originalPrint  = SchemaPrinter::doPrint($this->testSchema);
+        $ast            = Parser::parse($sdl);
+        $extendedSchema = SchemaExtender::extend($this->testSchema, $ast, $options);
+        $this->assertEquals(SchemaPrinter::doPrint($this->testSchema), $originalPrint);
```
Comment (simPod) [inline tests/Utils/SchemaExtenderTest.php]: can assertions be called statically where appropriate?

For ref see https://github.com/webonyx/graphql-php/pull/354#issuecomment-422900157

**Comment by Torsten85** [comment]
good enough? ;)

**Comment by vladar** [comment]
Great work!

File: src/Validator/DocumentValidator.php (hunk 3 above)
Comment (vladar) [reply src/Validator/DocumentValidator.php]: Actually, we don't use carriage returns anywhere, so `\n` is appropriate. PHP_EOL is not for our case. I'll revert it.

**Pull request merged**